### DISCUSSION TITLE
Is there a strategy regarding choice of "aria-hidden" vs. (for example) "aria-label"?

### DISCUSSION THREAD

**Discussion by gmerritt** [PrairieLearn/PrairieLearn#9371] [Q&A] [answered]
I've encountered some interface elements for which accessibility evaluation tools give a WCAG 2.1 level A warning about "Button missing a text alternative" (4.1.2 Name, Role, Value). Examples include the little question mark buttons for more info, such as on [a student's "Assessments" summary page](https://us.prairielearn.com/pl/course_instance/127369/assessments) (alongside entries in the "Available credit" column) and on [an instructors "Course Instances" page](https://us.prairielearn.com/pl/course/6387/course_admin/instances) (with the "Earliest Access Date" and "Latest Access Date" column headers).

In the PrairieLearn code at large, sometimes the use of "[aria-hidden](https://github.com/search?q=repo%3APrairieLearn%2FPrairieLearn%20aria-hidden&type=code)" for an element will prevent that element from being in contraction of the WCAG guidelines, while at other times "[aria-label](https://github.com/search?q=repo%3APrairieLearn%2FPrairieLearn+aria-label&type=code)" similarly prevents other elements from raising a flag.

Are there general practices or guidelines regarding when "aria-hidden" might be chosen vs. "aria-label"? I might consider that labels would be preferred in the examples above, as the related elements could be important for people accessing PL with ARIA-enabled technology, but I would like to know if there are presently any specific guidelines/conventions regarding how such elements are best handled for the PL context.

Thanks!

**Comment by nwalters512** [comment, marked as answer]
I don't view these as an either/or; they're complementary.

Pretty much all the instances I see of `aria-hidden` are for `<i>` elements inside buttons/links, which we intentionally don't want to show to screen readers. Wherever there's an icon, we should also have text describing alongside it, provided either via `aria-label` (in the case of a button with no text that's visible in the UI) or via just putting the text inside the button.

You've found two examples where this isn't working as intended. Our _intent_ was that the `title` attribute for these links would be read by screen readers, but it looks like Bootstrap is taking the title attribute's value and moving it to `data-original-title`, which screen readers can't pick up. In these cases, I think using `aria-label` on the link would be appropriate, and I would encourage you to file either issues or PRs to fix problems that your accessibility tooling identifies.

Aside: the fact that our automated accessibility testing doesn't complain about these is because that tooling runs against the raw HTML without executing JavaScript. It sees that there's a `title` attribute and so it doesn't complain. It might be useful for us to add a custom check that asserts that any popover triggers use `aria-label` if the trigger doesn't itself have any text.

**Reply by gmerritt** [reply]
Thank you, Nathan! That all makes sense.

Yes, I am getting some inklings of the subtleties that may be required to understand how pages are encoded and then built/served when making these kinds of evaluations in any modern dynamic-page-building content management system.

Ok, I will indeed have a go at some PRs in this area. Thanks again!
